source/MRViewer/MRSurfaceManipulationWidget.cpp

```diff
@@ -30,38 +30,28 @@
 #include "MRMesh/MRChangeSelectionAction.h"
 #include "MRMesh/MRObjectsAccess.h"
 #include "MRSceneCache.h"
+#include "MRMesh/MRAABBTreePoints.h"
+#include "MRMesh/MRPointsProject.h"
 
 namespace MR
 {
 
-void findSpaceDistancesAndCodirectedVerts( const Mesh& mesh, const PointOnFace& start, float range, const Vector3f& dir,
-                                             VertScalars& distances, VertBitSet& verts )
-{
-    MR_TIMER;
-
-    EnumNeihbourVertices e;
-    e.run( mesh.topology, mesh.getClosestVertex( start ), [&] ( VertId v )
-    {
-        const float dist = ( start.point - mesh.points[v] ).length();
-        const bool valid = ( dist <= range ) && ( dot( mesh.normal( v ), dir ) >= 0.f );
-        distances[v] = dist;
-        verts.set( v, valid );
-        return valid;
-    } );
-}
-
-void findSpaceDistancesAndVerts( const Mesh& mesh, const PointOnFace& start, float range, VertScalars& distances, VertBitSet& verts )
+void findSpaceDistancesAndVerts( const Mesh& mesh, const VertBitSet& start, float range, VertScalars& distances, VertBitSet& verts, bool codirected, const VertBitSet* untouchable )
 {
     // note! better update bitset for interesting verts than check for all verts from distances
     MR_TIMER;
+    auto tree = AABBTreePoints( mesh.points, start );
```

Comment:
Why not mesh.getAABBTreePoints?

Reply:
we only build tree of small number of `start` vertices, mesh.getAABBTreePoints() get tree from all points